lib/pbench/test/unit/server/query_apis/commons.py

```diff
@@ -111,43 +126,72 @@ def test_missing_json_object(self, client, server_config, pbench_token):
+                json=keys,
+            )
+            assert response.status_code == HTTPStatus.BAD_REQUEST
+            missing = [k for k in required_keys if k not in keys]
+            assert (
+                response.json.get("message")
+                == f"Missing required parameters: {','.join(missing)}"
+            )
+
+        parameter_items = self.cls_obj.schema.parameters.items()
+
         required_keys = [
             key
             for key, parameter in self.cls_obj.schema.parameters.items()
             if parameter.required
         ]
-        response = client.post(
-            server_config.rest_uri + self.pbench_endpoint,
-            headers={"Authorization": "Bearer " + pbench_token},
-            json=keys,
-        )
-        assert response.status_code == HTTPStatus.BAD_REQUEST
-        missing = [k for k in required_keys if k not in keys]
-        assert (
-            response.json.get("message")
-            == f"Missing required parameters: {','.join(missing)}"
-        )
+
+        all_combinations = []
+        for r in range(1, len(parameter_items) + 1):
+            combinations_object = list(itertools.combinations(parameter_items, r))
+            for item in combinations_object:
+                tmp_req_keys = []
+                for element in item:
+                    if element[1].required:
+                        tmp_req_keys.append(element[0])
+                if not sorted(tmp_req_keys) == sorted(required_keys):
+                    all_combinations.append(item)
+
+        for items in all_combinations:
+            keys = {}
+            for item in items:
+                if item[1].type == ParamType.ACCESS:
+                    keys[item[0]] = "public"
+                elif item[1].type == ParamType.DATE:
+                    keys[item[0]] = "2020"
+                else:
+                    keys[item[0]] = "foobar"
+
+            missing_key_helper(keys)
+
+        # Test for random non-existent key
+        missing_key_helper({"notakey": None})
 
     def test_bad_dates(self, client, server_config, user_ok, pbench_token):
         """
         Test behavior when a bad date string is given
         """
         parameter_types = [
-            parameter.type
-            for _, parameter in self.cls_obj.schema.parameters.items()
-            if parameter.required
+            parameter.type for _, parameter in self.cls_obj.schema.parameters.items()
         ]
-        if ParamType.DATE not in parameter_types or not all(
-            k in self.payload for k in ("start", "end")
+        if ParamType.DATE not in parameter_types or any(
+            k not in self.payload for k in ("start", "end")
         ):
-            pytest.skip("skipping the bad date test")
+            pytest.skip("skipping " + self.test_bad_dates.__name__)
 
         # Modify start and end time in the payload to make it look invalid
         self.payload["start"] = "2020-12"
```

Comment:
OK, well, I had in mind something more general; but I get that the silly Elasticsearch query URI kinda relies on start and end specifically. However, in that case, looking for `DATE` parameters doesn't really serve much purpose, unless you want to verify that the `start` and `end` parameters are both in the schema *and* have `DATE` type, which would be a bit safer although nobody's likely to use those names for anything else.

Reply:
OK, I'll second this request.  My motivation is that, if a request has either `"start"` **_or_** `"end"`, it should probably be considered legitimate.  That is, while we presumably don't do so currently, we _could_ support open-ended ranges.  So, having the test skip when we specify only one end of the range is probably overly restrictive.

From a practical perspective, I expect that Nikhil wants to skip the test if either `"start"` or `"end"` is missing because his code is going to overwrite those two fields.  But, it's easy enough to make the code skip any fields which don't exist...or, as Dave suggests, to have it iterate over the date fields which do exist, stomping on each one in sequence, repeating the request each time, which would be good and thorough (and general).